File: factorization_machine/data_loader.py
import numpy as np

def getUserWatchedCnt(dir="../item_recom/train_info.tsv"):
	stats_user = [0]*94317
	stats_item = [0]*99782
	with open(dir) as file:
		for line in file:
			(user, item, week, time, feat1, feat2, rating)=line.split('\t')
			stats_user[int(user)] += 1
			stats_item[int(item)] += 1
	return stats_user, stats_item

def loadData(dir, train_ratio=0.9):
	train, train_rating = [], []
	valid, valid_rating = [], []
	cnt = 0
	print "Start loading data..."
	stats_user, stats_item = getUserWatchedCnt(dir)
	with open(dir) as file:
		for line in file:
			if cnt % 10000 == 9999: 
				print "%d / 5974450 loaded\r"%(cnt+1),
			cnt += 1
			# if cnt == 100000: break
			(user, item, week, time, feat1, feat2, rating)=line.split('\t')
			if np.random.rand() < train_ratio:
				train.append(
						{"1_user_id": str(int(user)),
						 "2_item_id": str(int(item)),
						 "5_week":	str(int(week)),
						 "6_time":	str(int(time)),
						 "3_feat1":	str(int(feat1)),
						 "4_feat2":	str(int(feat2)),
						 "7_uwatched": 1.0/float(stats_user[int(user)]),
						 "8_iwatched": 1.0/float(stats_item[int(item)])
						})
				train_rating.append(float(rating))
			else:
				valid.append(
					{"1_user_id": str(int(user)),
					 "2_item_id": str(int(item)),
					 "5_week":	str(int(week)),
					 "6_time":	str(int(time)),
					 "3_feat1":	str(int(feat1)),
					 "4_feat2":	str(int(feat2)),
					 "7_uwatched": 1.0/float(stats_user[int(user)]),
					 "8_iwatched": 1.0/float(stats_item[int(item)])
					})
				valid_rating.append(float(rating))			

	return (train, np.array(train_rating)), (valid, np.array(valid_rating))

def loadTest(dir):
	test = []
	cnt = 0
	print "Start loading test data..."
	stats_user, stats_item = getUserWatchedCnt()
	with open(dir) as file:
		for line in file:
			if cnt % 10000 == 9999: 
				print "%d / 1524458 loaded\r"%(cnt+1),
			cnt += 1
			# if cnt == 100000: break
			(user, item, week, time, feat1, feat2)=line.split('\t')
			test.append(
				{"1_user_id": str(int(user)),
				 "2_item_id": str(int(item)),
				 "5_week":	str(int(week)),
				 "6_time":	str(int(time)),
				 "3_feat1":	str(int(feat1)),
				 "4_feat2":	str(int(feat2)),
				 "7_uwatched": 1.0/float(stats_user[int(user)]),
				 "8_iwatched": 1.0/float(stats_item[int(item)])
				})		

	return test

File: autoencoder/data_loader.lua
require "torch"
require "nnsparse"

-- Meta Class
Dataset = {trainV = {}, validV = {}, trainU = {}, validU = {}}

function Dataset:new(obj, dir, train_ratio)
	obj = obj or {}
	setmetatable(obj, self)
	self.__index = self
	self.trainV, self.validV, self.trainU, self.validU = self:load_data(dir)
	-- self.bias = self:unbias()

	return Dataset
end

function Dataset:load_data(dir)
	local trainV, validV, trainU, validU = {}, {}, {}, {}
	local trainFile = io.open(dir, "r")
	print("Start loading data")
	cnt, cnt_train, cnt_valid = 0, 0, 0

	for line in trainFile:lines() do
		local userIdStr, itemIdStr, weekStr, timeStr, feat1Str, feat2Str, ratingStr 
				= line:match('(%d+)\t(%d+)\t(%d+)\t(%d+)\t(%d+)\t(%d+)\t(%d%.?%d?)')
		local userId = tonumber(userIdStr)+1
		local itemId = tonumber(itemIdStr)+1
		local week = tonumber(weekStr)
		local time = tonumber(timeStr)
		local feat1 = tonumber(feat1Str)
		local feat2 = tonumber(feat2Str)
		local rating = tonumber(ratingStr)

		-- normalize the rating between interval [-1, 1]
		rating = (rating-2.5)/2.5
		
		if torch.uniform() < train_ratio then
			-- prepare item-based training set
			trainV[itemId] = trainV[itemId] or nnsparse.DynamicSparseTensor()
			trainV[itemId]:append(torch.Tensor{userId, rating})
			-- prepare user-based training set
			trainU[userId] = trainU[userId] or nnsparse.DynamicSparseTensor()
			trainU[userId]:append(torch.Tensor{itemId, rating})
			cnt_train = cnt_train + 1
		else
			-- prepare item-based validation set
			validV[itemId] = validV[itemId] or nnsparse.DynamicSparseTensor()
			validV[itemId]:append(torch.Tensor{userId, rating})
			-- prepare user-based validation set
			validU[userId] = validU[userId] or nnsparse.DynamicSparseTensor()
			validU[userId]:append(torch.Tensor{itemId, rating})
			cnt_valid = cnt_valid + 1
		end

		if cnt % 10000 == 9999 then xlua.progress(cnt+1, 5974450) end

		cnt = cnt + 1
		-- if cnt > 100000 then break end
	end

	for k, train_item in pairs(trainV) do trainV[k] = train_item:build():ssortByIndex() end
	for k, valid_item in pairs(validV) do validV[k] = valid_item:build():ssortByIndex() end

	for k, train_item in pairs(trainU) do trainU[k] = train_item:build():ssortByIndex() end
	for k, valid_item in pairs(validU) do validU[k] = valid_item:build():ssortByIndex() end

	print("Finish data loading!")
	print(cnt_train, "records are in training set")
	print(cnt_valid, "records are in validation set")

	return trainV, validV, trainU, validU
end

-- function Dataset:unbias()
-- 	bias = {}
-- 	for k, train_item in pairs(self.train) do
-- 		local mean = train_item[{{}, 2}]:mean()
-- 		self.train[k][{{}, 2}]:add(-mean)
		
-- 		if self.valid[k] ~= nil and self.valid[k].curSize == nil then
-- 			self.valid[k][{{}, 2}]:add(-mean)
-- 		end
-- 		bias[k] = mean
-- 		if k % 1000 == 999 then xlua.progress(k, 99782) end
-- 	end
-- 	return bias
-- end

return Dataset

-- dataset = Dataset:new(nil, "item_recom/train_info.tsv", 0.9)
